docs/design/request-compression-trait.md

```diff
@@ -0,0 +1,332 @@
+            supportedCompressionAlgorithms.find { supported ->
+                requestedCompressionAlgorithms.find { supported.id == it } != null
+            }?.let { algorithm ->
+
+                // Attempt compression
+                val request = context.protocolRequest.toBuilder()
+                request.body = request.body
+                // TODO: Write compression
+                request.headers.append("Content-Encoding", algorithm.id)
+                return request.build()
+            }
+        }
+        return context.protocolRequest
+    }
+}
+```
+
+### Testing
+Testing is done using both unit tests and Smithy HTTP protocol compliance tests. The SDK unit tests are used to verify behaviour and ensure it doesn't drift over time. AWS HTTP protocol tests are used to test the behavior in a more integrated way.
+
+### Appendix
+
+#### Codegen
+Middleware registration has the following steps:
+
+* A Kotlin integration will look at a service model
+* Determine if the service has an operation with the `requestCompression` trait
+* If so middleware is registered  for the service
+* The middleware will look at each operation in the service individually
+* The middleware will then determine if an operation has the `requestCompression` trait
+* If so then it'll add an interceptor to the operation
+
+```kotlin
+/**
+ * Middleware that enables compression of payloads for HTTP requests
+ */
+class RequestCompressionTrait : KotlinIntegration {
+
+    // Will determine if middleware is registered for service
+    override fun enabledForService(model: Model, settings: KotlinSettings): Boolean = model
+        .shapes<OperationShape>()
+        .any { it.hasTrait<RequestCompressionTrait>() }
+
+    // Registers middleware
+    override fun customizeMiddleware(
+        ctx: ProtocolGenerator.GenerationContext,
+        resolved: List<ProtocolMiddleware>
+    ): List<ProtocolMiddleware> = resolved + requestCompressionTraitMiddleware
+
+    // TODO: Use `additionalServiceConfigProps` to register `compressionAlgorithms`
+    // TODO: Register function for client config to support user supplied compression algorithms (`addCompressionAlgorithms`)
+
+    // Middleware
+    private val requestCompressionTraitMiddleware = object : ProtocolMiddleware {
+        private val interceptorSymbol = RuntimeTypes.HttpClient.Interceptors.RequestCompressionTraitInterceptor
+        override val name: String = "RequestCompressionTrait"
+
+        // Will add interceptor to operation(s) in service with `requestCompression` trait 
+        override fun render(ctx: ProtocolGenerator.GenerationContext, op: OperationShape, writer: KotlinWriter) {
+            op.getTrait<RequestCompressionTrait>()?.let { trait ->
+                val requestedCompressionAlgorithms = trait.encodings
+
+                writer.withBlock(
+                    "if (config.disableRequestCompression == false || config.disableRequestCompression == null) {",
```

Comment:
Can `disableRequestCompression` be null? We should make it non-nullable in codegen because it has a default, which would simplify its use.